File: src/main/java/jvmusin/customaggregations/ContentsValidator.java
package jvmusin.customaggregations;

import org.jetbrains.annotations.NotNull;

import static java.text.MessageFormat.format;

/**
 * Contents validator.
 *
 * <p>Used to validate {@link Contents}.
 *
 * <p>Basically it checks that the first {@link Header} has level {@code 1}
 * and it's possible to build a single table of contents from the given contents instance.
 */
public class ContentsValidator {

    /**
     * Validates the {@code contents}.
     *
     * <p>Throws {@link ContentsInvalidException} if:
     * <ul>
     *     <li>the first header has indentation level other than `1`, or</li>
     *     <li>indentation level increases by more than `1` in some pair of two consecutive headers, or</li>
     *     <li>indentation level of some header is less than `1`.</li>
     * </ul>
     *
     * @param contents {@link Contents} to validate.
     * @throws ContentsInvalidException if the {@code contents} is invalid.
     */
    public void validate(@NotNull Contents contents) {
        if (!contents.isEmpty() && contents.get(0).getLevel() != 1) {
            throw new ContentsInvalidException(
                    "First header of the document should have indentation level 1 (a single hash symbol): " + contents.get(0)
            );
        }
        for (int i = 1; i < contents.size(); i++) {
            Header prev = contents.get(i - 1);
            Header cur = contents.get(i);
            if (cur.getLevel() < 1 || cur.getLevel() > prev.getLevel() + 1) {
                throw new ContentsInvalidException(
                        format("Headers '{0}' and '{1}' can't go after each other", prev, cur)
                );
            }
        }
    }
}

File: src/main/kotlin/jvmusin/customaggregations/ContentsValidator.kt
package jvmusin.customaggregations

/**
 * Contents validator.
 *
 * Used to validate [Contents].
 *
 * Basically it checks that the first [Header] has level `1`
 * and it's possible to build a single table of contents from the given [Contents] instance.
 */
class ContentsValidator {

    /**
     * Validates the [contents].
     *
     * Throws [ContentsInvalidException] if:
     * * the first header has indentation level other than `1`, or
     * * indentation level increases by more than `1` in some pair of two consecutive headers, or
     * * indentation level of some header is less than `1`.
     *
     * @param contents [Contents] to validate.
     * @throws ContentsInvalidException if the [contents] is invalid.
     */
    fun validate(contents: Contents) {
        if (contents.isNotEmpty() && contents.first().level != 1) {
            throw ContentsInvalidException(
                "First header of the document should have indentation level 1 (a single hash symbol): " +
                        "'${contents.first()}'"
            )
        }
        for ((prev, cur) in contents.zipWithNext()) {
            if (cur.level !in 1..prev.level + 1) {
                throw ContentsInvalidException("Headers '$prev' and '$cur' can't go after each other")
            }
        }
    }
}
